benchmark/benchmark-atomic.cpp

```diff
@@ -0,0 +1,220 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-24, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+// RAJA/RAJA.hpp cannot be included here because the include logic will
+// default all atomic implementations to a desul backend.
+#include "RAJA/policy/openmp/atomic.hpp"
+#include "RAJA/RAJA.hpp"
+#include "RAJA/policy/openmp/policy.hpp"
+
+#include <type_traits>
+#include <iostream>
+#include <sstream>
+#include <string>
+
+
+/// Conditional compilation for CUDA benchmarks.
+#if defined RAJA_ENABLE_CUDA
+#include "RAJA/policy/cuda.hpp"
+#include "RAJA/policy/cuda/atomic.hpp"
+
+template<int BLOCK_SZ>
+struct ExecPolicyGPU {
+    using policy = RAJA::cuda_exec<BLOCK_SZ, false>;
+    static std::string PolicyName() {
+        std::stringstream ss;
+        ss << "CUDA execution with block size " << BLOCK_SZ;
+        return ss.str();
+    }
+};
+
+struct GPUAtomic {
+    using policy = RAJA::policy::cuda::cuda_atomic;
+};
+
+#elif defined RAJA_ENABLE_HIP
+#include "RAJA/policy/hip.hpp"
+#include "RAJA/policy/hip/atomic.hpp"
+
+template<int M>
+struct IsGPU<RAJA::hip_exec<M>> : public std::true_type {};
+
+template<int BLOCK_SZ>
+struct ExecPolicyGPU {
+    using policy = RAJA::hip_exec<BLOCK_SZ, false>;
+    static std::string PolicyName() {
+        std::stringstream ss;
+        ss << "HIP execution with block size " << BLOCK_SZ;
+        return ss.str();
+    }
+};
+
+struct GPUAtomic {
+    using policy = RAJA::policy::hip::hip_atomic;
+};
+
+#endif
+
+#include "desul/atomics.hpp"
+#include "RAJA/util/Timer.hpp"
+
+#define BLOCK_SZ 256
+#define INDENT "  "
+using raja_default_desul_order = desul::MemoryOrderRelaxed;
+using raja_default_desul_scope = desul::MemoryScopeDevice;
+
+// Desul atomics have a different signature than RAJA's built in ops. The following code provides some
```

Comment:
This is a concern I have. We will be running and timing desul atomics without going through the RAJA API, which is what users would do, at least until we support only desul atomics (if and when we do that is not clear). Specifically, we have removed part of the abstraction layer by not going through the RAJA interface and compilers may optimize differently as a result.